experimental/ast/path.go

```diff
@@ -0,0 +1,210 @@
+// the nil token.
+func (p Path) AsIdent() Token {
+	var tok Token
+	var count int
+	p.Components(func(c PathComponent) bool {
+		if count > 0 {
+			tok = Token{}
+			return false
+		}
+
+		if c.Separator().Nil() {
+			tok = c.AsIdent()
+		}
+
+		count++
+		return true
+	})
+	return tok
+}
+
+// AsBuiltin returns the builtin that this path represents.
+//
+// If this path does not represent a builtin, returns [BuiltinUnknown].
+func (p Path) AsBuiltin() Builtin {
+	return BuiltinByName(p.AsIdent().Text())
+}
+
+// Span implements [Spanner] for Path.
+func (p Path) Span() Span {
+	return JoinSpans(p.raw[0].With(p), p.raw[1].With(p))
+}
+
+// Components is an [iter.Seq] that ranges over each component in this path. Specifically,
+// it yields the (nilable) dot that precedes the component, and the identifier token.
+func (p Path) Components(yield func(PathComponent) bool) {
+	if p.Nil() {
+		return
+	}
+
+	first := p.raw[0].With(p)
+	if synth := first.synthetic(); synth != nil {
+		panic("synthetic paths are not implemented yet")
+	}
+
+	cursor := Cursor{
+		withContext: p.withContext,
+		start:       p.raw[0],
+		end:         p.raw[1] + 1, // Remember, Cursor.end is exclusive!
+	}
+
+	var sep Token
+	var broken bool
+	cursor.Iter(func(tok Token) bool {
+		if tok.Text() == "." || tok.Text() == "/" {
+			if !sep.Nil() {
+				// Uh-oh, empty path component!
+				if !yield(PathComponent{sep, Token{}}) {
+					broken = true
+					return false
+				}
+			}
+			sep = tok
+			return true
+		}
+
+		if !yield(PathComponent{sep, tok}) {
+			broken = true
+			return false
+		}
+		sep = Token{}
+		return true
+	})
+	if !broken && !sep.Nil() {
+		yield(PathComponent{sep, Token{}})
+	}
+}
+
+// PathComponent is a piece of a path. This is either an identifier or a nested path
+// (for an extension name).
+type PathComponent struct {
+	separator, name Token
+}
+
+// Separator is the token that separates this component fromm the previous one, if
+// any. This may be a dot or a slash.
+func (p PathComponent) Separator() Token {
+	return p.separator
+}
+
+// Name is the token that represents this component's name. THis is either an
+// identifier or a (...) token containing a path.
+func (p PathComponent) Name() Token {
+	return p.name
+}
+
+// Returns whether this is an empty path component. Such components are not allowed
+// in the grammar but may occur in invalid inputs nonetheless.
+func (p PathComponent) IsEmpty() bool {
+	return p.Name().Nil()
+}
+
+// IsExtension returns whether this path component is an extension component, i.e.
+// (a.b.c).
+func (p PathComponent) IsExtension() bool {
+	return p.Name().Kind() == TokenPunct
+}
+
+// AsExtension returns the Path inside of this path component, if it is an extension
+// path component.
+func (p PathComponent) AsExtension() Path {
+	if !p.IsExtension() {
+		return Path{}
+	}
+
+	// If this is a synthetic token, its children are already precisely a path,
```

Comment:
What exactly is a synthetic token? Where will they come from? Is this only for tree mutations? From some context elsewhere in this PR, looks like maybe the lexer will also generate synthetic tokens, to "merge" adjacent string literals into a single string token. Is that right? Anything else?

Reply:
I explain this in detail in the updated doc.go